README.md
# PHP Contrast Tools

[![Latest Version on Packagist](https://img.shields.io/packagist/v/breadthe/php-contrast.svg?style=flat-square)](https://packagist.org/packages/breadthe/php-contrast)
[![Build Status](https://img.shields.io/travis/breadthe/php-contrast/master.svg?style=flat-square)](https://travis-ci.org/breadthe/php-contrast)
[![Quality Score](https://img.shields.io/scrutinizer/g/breadthe/php-contrast.svg?style=flat-square)](https://scrutinizer-ci.com/g/breadthe/php-contrast)
[![Total Downloads](https://img.shields.io/packagist/dt/breadthe/php-contrast.svg?style=flat-square)](https://packagist.org/packages/breadthe/php-contrast)

Provides various utilities for working with color contrast.

The primary purpose is to facilitate easy generation of accessible (minimum 3.0 contrast ratio) hex color pairs. A variation of these tools restricts the generated colors to [TailwindCSS's](https://tailwindcss.com/) default palette.

I've created a [demo page](https://mumu.pw/contrast) to show how color pairs can be generated by this package. 

## Installation

You can install the package via composer:

```bash
composer require breadthe/php-contrast
```

## Usage

Import the class.

```php
use Breadthe\PhpContrast\HexColor; // factory

use Breadthe\PhpContrast\HexColorPair; // hex pair utilities

use Breadthe\PhpContrast\TailwindColor; // Tailwind color pair utilities
```

## Check the contrast ratio between 2 colors

```php
$hexColorPair = HexColorPair::make(HexColor::make('000000'), HexColor::make('ffffff'));
$hexColorPair->ratio; // 21
$hexColorPair->fg; // '#000000'
$hexColorPair->bg; // '#ffffff'
```

## Get a random color pair (with the resulting ratio), with minimum 3:1 contrast ratio

```php
$hexColorPair = HexColorPair::random();
$hexColorPair->ratio; // 3.8
$hexColorPair->fg->hex; // '#36097e'
$hexColorPair->fg->name; // null
$hexColorPair->bg->hex; // '#ed4847'
$hexColorPair->bg->name; // null
```

## Get a random color pair (with the resulting ratio), with minimum specified contrast ratio (but no less than 3:1)

**⚠️ Warning** For performance reasons, the minimum requested contrast ratio is capped at 4.5, although the generated pairs can go up to the theoretical maximum 21:1 ratio.

**⚠️ Caution** When chaining with `minContrast()`, make sure to use `getRandom()` instead of `random()`.

```php
$hexColorPair = HexColorPair::minContrast(4.5)->getRandom();
$hexColorPair->ratio; // 7.6
$hexColorPair->fg->hex; // '#0c402f'
$hexColorPair->fg->name; // null
$hexColorPair->bg->hex; // '#badd73'
$hexColorPair->bg->name; // null
```

## Get a random accessible sibling for the given color, with minimum specified contrast ratio (but no less than 3:1)

```php
// Minimum 3:1 contrast ratio
HexColorPair::sibling('000000'); // [21, '000000', 'ffffff']

// Minimum specified contrast ratio (no less than 3:1)
HexColorPair::minContrast(4.5)->getSibling('000000'); // [21, '000000', 'ffffff']
```

## Generate a random TailwindCSS color

```php
$twColor = TailwindColor::random();
$twColor->hex; // '#e2e8f0'
$twColor->name; // 'gray-300'
```

## Generate a pair of random accessible TailwindCSS colors

```php
$twColorpair = TailwindColor::randomPair();
$twColorpair->ratio; // 3.3
$twColorpair->fg->hex; // '#63b3ed'
$twColorpair->fg->name; // 'blue-400'
$twColorpair->bg->hex; // '#4a5568'
$twColorpair->bg->name; // 'green-700'
```

## Generate a pair of random accessible TailwindCSS colors, with minimum specified contrast ratio (but no less than 3:1)

**⚠️ Warning** For performance reasons, the minimum requested contrast ratio is capped at 4.5, although the generated pairs can go up to the theoretical maximum 21:1 ratio.

**⚠️ Caution** When chaining with `minContrast()`, make sure to use `getRandomPair()` instead of `randomPair()`.

```php
$twColorpair = TailwindColor::minContrast(4.5)->getRandomPair();
$twColorpair->ratio; // 7.0
$twColorpair->fg->hex; // '#faf5ff'
$twColorpair->fg->name; // 'purple-100'
$twColorpair->bg->hex; // '#9b2c2c'
$twColorpair->bg->name; // 'red-800'
```

## Merge the default Tailwind colors with a custom palette

You may extend the default Tailwind colors with your own custom palette. Here's an example of how to import a custom palette from a JSON file.

```json
{
    "background": "#FFFFFF",
    "headline": "#1f1235",
    "sub-headline": "#1b1425",
    "button": "#ff6e6c",
    "button-text": "#1f1235"
}
```

```php
$customPalette = json_decode(file_get_contents('custom-palette.json'), true);

$colors = TailwindColor::merge($customPalette)->getColors();
```

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email omigoshdev@protonmail.com instead of using the issue tracker.

## Credits

- [Omigosh Dev](https://github.com/breadthe)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## PHP Package Boilerplate

This package was generated using the [PHP Package Boilerplate](https://laravelpackageboilerplate.com).


CHANGELOG.md
# Changelog

All notable changes to `php-contrast` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/HexColorTest.php
<?php

namespace Breadthe\PhpContrast\Tests;

use Breadthe\PhpContrast\HexColor;
use PHPUnit\Framework\TestCase;

class HexColorTest extends TestCase
{
    /** @test */
    public function a_hex_color_is_created_correctly()
    {
        $hexColor = HexColor::make('abc');
        self::assertNull($hexColor->name);
        self::assertEquals('#aabbcc', $hexColor->hex);

        $hexColor = HexColor::make('#abc');
        self::assertNull($hexColor->name);
        self::assertEquals('#aabbcc', $hexColor->hex);

        $hexColor = HexColor::make('abcdef');
        self::assertNull($hexColor->name);
        self::assertEquals('#abcdef', $hexColor->hex);

        $hexColor = HexColor::make('#abcdef');
        self::assertNull($hexColor->name);
        self::assertEquals('#abcdef', $hexColor->hex);

        $name = 'custom-color-name';
        $hexColor = HexColor::make('#abcdef', $name);
        self::assertEquals($name, $hexColor->name);
        self::assertEquals('#abcdef', $hexColor->hex);
    }

    /** @test */
    public function a_random_hex_color_is_generated_correctly()
    {
        $randomHexColor = HexColor::random();

        self::assertRegExp('/^#[0-9a-f]{6}$/i', $randomHexColor->hex);
        self::assertNull($randomHexColor->name);
    }
}


tests/TailwindTest.php
<?php

namespace Breadthe\PhpContrast\Tests;

use Breadthe\PhpContrast\HexColor;
use Breadthe\PhpContrast\HexColorPair;
use Breadthe\PhpContrast\TailwindColor;
use PHPUnit\Framework\TestCase;

class TailwindTest extends TestCase
{
    /** @test */
    public function it_generates_the_correct_object_types()
    {
        $twColors = TailwindColor::colors();
        self::assertTrue(is_array($twColors));

        $randomTwColor = TailwindColor::random();
        self::assertTrue(is_a($randomTwColor, HexColor::class));

        $randomPair = TailwindColor::randomPair();
        self::assertTrue(is_a($randomPair, HexColorPair::class));
        self::assertTrue(is_a($randomPair->fg, HexColor::class));
        self::assertTrue(is_a($randomPair->bg, HexColor::class));
    }

    /** @test */
    public function it_returns_random_color_pairs_of_minimum_3_to_1_ratio()
    {
        $minRatio = 3;

        $randomPair = TailwindColor::randomPair();
        self::assertGreaterThanOrEqual($minRatio, $randomPair->ratio);

        // Color pairs with a desired min ratio < 3 will default to 3
        $desiredMinRatio = 2;
        $randomPair = TailwindColor::minContrast($desiredMinRatio)->getRandomPair();
        self::assertGreaterThanOrEqual($minRatio, $randomPair->ratio);

        // Color pairs are generated with the minimum specified ratio
        $desiredMinRatio = 4.5;
        $randomPair = TailwindColor::minContrast($desiredMinRatio)->getRandomPair();
        self::assertGreaterThanOrEqual($minRatio, $randomPair->ratio);
    }

    /** @test */
    public function it_can_merge_default_and_custom_colors()
    {
        $customPalette = json_decode(file_get_contents(__DIR__.'/../stubs/custom-palette.json'), true);

        $colors = TailwindColor::merge($customPalette)->getColors();
        $customPalette = collect($customPalette)
            ->map(function ($hex, $name) {
                return HexColor::make($hex, $name);
            })
            ->flatten()
            ->toArray();

        $randomCustomPaletteElement = $customPalette[rand(0, count($customPalette) - 1)];

        self::assertContains($randomCustomPaletteElement, $colors, false, true, false);
    }
}


tests/HexColorPairTest.php
<?php

namespace Breadthe\PhpContrast\Tests;

use Breadthe\PhpContrast\HexColor;
use Breadthe\PhpContrast\HexColorPair;
use PHPUnit\Framework\TestCase;

class HexColorPairTest extends TestCase
{
    /**
     * @test
     * @dataProvider provideContrastData
     */
    public function it_returns_the_correct_contrast_ratio(float $contrastRatio, array $colorPair)
    {
        $fg = HexColor::make($colorPair[0]);
        $bg = HexColor::make($colorPair[1]);
        $pair = HexColorPair::make($fg, $bg);

        // Test the property
        self::assertEquals($contrastRatio, $pair->ratio);

        // Test the static constructor
        self::assertEquals($contrastRatio, HexColorPair::ratio($fg, $bg));
    }

    /**
     * @test
     *
     * For more assurance, call the following assertions inside a loop
     */
    public function it_returns_random_color_pairs_of_minimum_3_to_1_ratio()
    {
        $minRatio = 3;

        $pair = HexColorPair::random();
        self::assertGreaterThanOrEqual($minRatio, $pair->ratio);

        // Color pairs with a desired min ratio < 3 will default to 3
        $desiredMinRatio = 2;
        $pair = HexColorPair::minContrast($desiredMinRatio)->getRandom();
        self::assertGreaterThanOrEqual($minRatio, $pair->ratio);

        // Color pairs are generated with the minimum specified ratio
        $desiredMinRatio = 4.5;
        $pair = HexColorPair::minContrast($desiredMinRatio)->getRandom();
        self::assertGreaterThanOrEqual($desiredMinRatio, $pair->ratio);
    }

    /** @test */
    public function a_sibling_of_minimum_contrast_is_generated_correctly()
    {
        $minRatio = 3;
        $hexColor = '#cccccc';

        // Sibling is generated with a minimum contrast ratio of 3.0
        $sibling = HexColorPair::sibling($hexColor);
        $pair = HexColorPair::make(HexColor::make($hexColor), $sibling);
        self::assertGreaterThanOrEqual($minRatio, $pair->ratio);

        // Sibling is generated with the minimum specified ratio
        $desiredMinRatio = 4.5;
        $sibling = HexColorPair::minContrast($desiredMinRatio)->getSibling($hexColor);
        $pair = HexColorPair::make(HexColor::make($hexColor), $sibling);
        self::assertGreaterThanOrEqual($desiredMinRatio, $pair->ratio);
    }

    public function provideContrastData()
    {
        yield 'maximum contrast' => [
            21,
            ['000', 'fff'],
        ];
        yield 'no contrast - black on black' => [
            1,
            ['000', '000'],
        ];
        yield 'no contrast - white on white' => [
            1,
            ['fff', 'fff'],
        ];
        yield 'high contrast' => [
            18.4,
            ['300', 'fff'],
        ];
        yield 'low contrast' => [
            1.1,
            ['300', '000'],
        ];
    }
}


src/HexColor.php
<?php

namespace Breadthe\PhpContrast;

class HexColor implements Color
{
    public $hex;
    public $name;

    public function __construct(string $hex, string $name = null)
    {
        $this->hex = $this->normalize($hex);
        $this->name = $name;
    }

    public static function make(string $hex, string $name = null): self
    {
        return new static($hex, $name);
    }

    public static function random(): self
    {
        return new static(self::randomColor());
    }

    /**
     * '#abc' or 'abc' => '#aabbcc'
     * '#abcdef' or 'abcdef' => '#abcdef'.
     */
    protected function normalize(string $hexColor)
    {
        // '#abcdef' or 'abcdef'
        if (preg_match('/^\s*#?([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})\s*$/i', $hexColor, $matches)) {
            return sprintf('#%s%s%s', $matches[1], $matches[2], $matches[3]);
        }

        // '#abc' or 'abc'
        if (preg_match('/^\s*#?([0-9a-f])([0-9a-f])([0-9a-f])\s*$/i', $hexColor, $matches)) {
            return sprintf('#%s%s%s', str_repeat($matches[1], 2), str_repeat($matches[2], 2),
                str_repeat($matches[3], 2));
        }

        // TODO throw
    }

    /**
     * @see https://stackoverflow.com/questions/5614530/generating-a-random-hex-color-code-with-php
     */
    protected static function randomColor()
    {
        return static::randomColorPart().static::randomColorPart().static::randomColorPart();
    }

    protected static function randomColorPart()
    {
        return str_pad(dechex(mt_rand(0, 255)), 2, '0', STR_PAD_LEFT);
    }
}


src/TailwindColor.php
<?php

namespace Breadthe\PhpContrast;

use Tightenco\Collect\Support\Collection;

class TailwindColor implements Color
{
    protected $colors;

    protected $minRatio;

    public function __construct()
    {
        $this->colors = $this->getDefaultTailwindColors();
    }

    /**
     * Returns all the default Tailwind colors.
     */
    public static function colors(): array
    {
        return (new static)->colors->toArray();
    }

    public function getColors(): array
    {
        return $this->colors->toArray();
    }

    /**
     * Returns a random Tailwind color.
     */
    public static function random(): HexColor
    {
        $tailwindColors = (new static)->colors;

        return $tailwindColors[rand(0, count($tailwindColors) - 1)];
    }

    public static function minContrast($ratio): self
    {
        if ($ratio < HexColorPair::MIN_RATIO) {
            $ratio = HexColorPair::MIN_RATIO;
        }

        if ($ratio > HexColorPair::MAX_RATIO) {
            $ratio = HexColorPair::MAX_RATIO;
        }

        $twColor = new static;
        $twColor->minRatio = $ratio;

        return $twColor;
    }

    public static function merge(array $customPalette): self
    {
        $twColor = new static;

        $customPalette = collect($customPalette)
            ->map(function ($hex, $name) {
                return HexColor::make($hex, $name);
            });

        $twColor->colors = $twColor->colors->merge($customPalette->flatten());

        return $twColor;
    }

    /**
     * Returns a random pair of accessible (min. contrast 3:1) Tailwind colors.
     */
    public static function randomPair(): HexColorPair
    {
        return (new static)->getRandomPair();
    }

    public function getRandomPair(): HexColorPair
    {
        $minRatio = $this->minRatio ?? HexColorPair::MIN_RATIO;

        $fg = static::random();

        return $this->getRandomPairMinRatio($fg, $minRatio);
    }

    protected function getRandomPairMinRatio($fg, $minRatio): HexColorPair
    {
        do {
            $bg = static::random();
            $pair = HexColorPair::make($fg, $bg);
        } while ($pair->ratio <= $minRatio);

        return $pair;
    }

    protected function getDefaultTailwindColors(): Collection
    {
        $twColorsArr = $this->parseDefaultTailwindColors();

        return $this->flatten($twColorsArr['colors']);
    }

    protected function flatten(array $twColorsArr): Collection
    {
        $colors = new Collection();

        collect($twColorsArr)
            ->each(function ($color, $label) use ($colors) {
                if (is_array($color)) {
                    foreach ($color as $shade => $value) {
                        $colors->push(HexColor::make($value, "{$label}-{$shade}"));
                    }
                } else {
                    $colors->push(HexColor::make($color, $label));
                }
            });

        return $colors;
    }

    protected function parseDefaultTailwindColors()
    {
        return json_decode(file_get_contents(__DIR__.'/../stubs/twcolors.json'), true);
    }
}


src/HexColorPair.php
<?php

namespace Breadthe\PhpContrast;

class HexColorPair
{
    const MIN_RATIO = 3.0;
    const MAX_RATIO = 4.5; // hard cap on the max requested ratio, to mitigate performance issues

    public $ratio; // computed contrast ratio between fg : bg

    // Calling these foreground/background instead of a more generic "color 1"/"color 2"
    // Doesn't really matter, they can be used interchangeably
    public $fg; // foreground (text) color
    public $bg; // background color

    protected $minRatio;

    public function __construct(HexColor $fg = null, HexColor $bg = null)
    {
        $this->fg = $fg;
        $this->bg = $bg;
        $this->ratio = static::calculateRatio($fg, $bg);
    }

    public static function make(HexColor $fg, HexColor $bg)
    {
        return new static($fg, $bg);
    }

    public static function ratio(HexColor $fg, HexColor $bg)
    {
        return (new static($fg, $bg))->ratio;
    }

    public static function random(): self
    {
        return (new static)->getRandom();
    }

    public static function sibling(string $hexValue): HexColor
    {
        return (new static)->getSibling($hexValue);
    }

    public static function minContrast($ratio): self
    {
        if ($ratio < self::MIN_RATIO) {
            $ratio = self::MIN_RATIO;
        }

        if ($ratio > self::MAX_RATIO) {
            $ratio = self::MAX_RATIO;
        }

        $pair = new static;
        $pair->minRatio = $ratio;

        return $pair;
    }

    public function getRandom(): self
    {
        $minRatio = $this->minRatio ?? self::MIN_RATIO;

        $fg = HexColor::random();

        return $this->getRandomPairMinRatio($fg, $minRatio);
    }

    public function getSibling(string $hexValue): HexColor
    {
        $minRatio = $this->minRatio ?? self::MIN_RATIO;

        $pair = $this->getRandomPairMinRatio(HexColor::make($hexValue), $minRatio);

        return $pair->bg;
    }

    protected static function calculateRatio(HexColor $fg = null, HexColor $bg = null)
    {
        if (! $fg || ! $bg) {
            return;
        }

        $fgLuminance = static::luminance($fg);
        $bgLuminance = static::luminance($bg);

        return round((max($fgLuminance, $bgLuminance) + 0.05) / (min($fgLuminance, $bgLuminance) + 0.05) * 10) / 10;
    }

    protected static function luminance(HexColor $color)
    {
        [$rHex, $gHex, $bHex] = static::rgbHexChannels($color);

        // Get decimal values
        $r8bit = base_convert($rHex, 16, 10);
        $g8bit = base_convert($gHex, 16, 10);
        $b8bit = base_convert($bHex, 16, 10);

        // Get sRGB values
        $rSrgb = $r8bit / 255;
        $gSrgb = $g8bit / 255;
        $bSrgb = $b8bit / 255;

        // Calculate luminance
        $r = ($rSrgb <= 0.03928) ? $rSrgb / 12.92 : pow((($rSrgb + 0.055) / 1.055), 2.4);
        $g = ($gSrgb <= 0.03928) ? $gSrgb / 12.92 : pow((($gSrgb + 0.055) / 1.055), 2.4);
        $b = ($bSrgb <= 0.03928) ? $bSrgb / 12.92 : pow((($bSrgb + 0.055) / 1.055), 2.4);

        return 0.2126 * $r + 0.7152 * $g + 0.0722 * $b;
    }

    /**
     * '#abcdef' => ['ab', 'cd', 'ef'].
     */
    protected static function rgbHexChannels(HexColor $hexColor): array
    {
        $hex = substr($hexColor->hex, 1); // strip the #

        return [
            substr($hex, 0, 2),
            substr($hex, 2, 2),
            substr($hex, 4, 2),
        ];
    }

    protected function getRandomPairMinRatio($fg, $minRatio): self
    {
        do {
            $bg = HexColor::random();
            $pair = new static($fg, $bg);
        } while ($pair->ratio <= $minRatio);

        return $pair;
    }
}
